### ISSUE TITLE
[ENH] Integration of LTSF-Linear Deep Learning Forecasters

### ISSUE THREAD

**Issue by luca-miniati** [sktime/sktime#4939]
This issue serves as an umbrella issue for integrating networks from LTSF-Linear. Deep learning has proven to be an effective way to predict time series data. To expand this type of forecasting in sktime, LTSF-Linear is a good place to start.

LTSF-Linear is a Pytorch implementation of the paper, [“Are Transformers Effective for Time Series Forecasting?”](https://arxiv.org/pdf/2205.13504.pdf). I am writing notebooks to reproduce the results, and will link them to this issue when completed.

Forecasters to add:

 - [ ] Linear Family https://github.com/cure-lab/LTSF-Linear/tree/main/models#ltsf-linear-family

   - [ ] Linear
   - [ ] DLinear
   - [ ] NLinear  
 - [ ] Transformer (NeuIPS 2017) https://github.com/thuml/Autoformer
 - [ ] Informer (AAAI 2021 Best paper) https://github.com/thuml/Autoformer
 - [ ] Autoformer (NeuIPS 2021) https://github.com/thuml/Autoformer
 - [ ] Pyraformer (ICLR 2022 Oral) https://github.com/alipay/Pyraformer
 - [ ] FEDformer (ICML 2022) https://github.com/MAZiqing/FEDformer

Currently, a BaseDeepForecasterPyTorch class is being developed, meant to bridge the gap between BaseForecaster and PyTorch-based networks. This base class will implement the pytorch backend of listed models.

Original authors of LTSF-Linear: @mixiancmx, @ailingzengzzz, @shaido987

**Comment by fkiraly** [comment]
FYI @kcc-lion - I think forecasting and neural networks might be of interest to you?

We've just started the summer programme and one of the workstreams is forecasting/deep learning, see here: https://github.com/sktime/sktime/projects/33

**Comment by ilkersigirci** [comment]
For integrating pytorch models, IMHO it is better to use [skorch](https://skorch.readthedocs.io/en/stable/user/quickstart.html) with sktime [YfromX](https://www.sktime.net/en/latest/api_reference/auto_generated/sktime.forecasting.compose.YfromX.html) wrapper. I have implemented example code [here](https://gist.github.com/ilkersigirci/d9add1d404bfea03913f95ae18bb0a73). Currently, grid search seems to be broken but I think it is easier to quickly fix it, rather than implementing hole new BaseClass for pytorch.

**Comment by fkiraly** [comment]
> Currently, grid search seems to be broken

Is it? Can you elaborate, please, @ilkersigirci? Broken how exactly?
The tests all seem to run.

**Comment by ilkersigirci** [comment]
Sorry for the misunderstanding. GridSearch is working fine with all sktime native forecasters. But, it gives error when I tried to use pytorch models with skorch. 
I first wrap my pytorch model with my customskorchregressor and then wrap again with `YfromX`. All steps can be found the notebook I shared on the above. When I try to run ForecastingGridSearch, it breaks when get_params. If requested, I can paste the error. 
BTW, it would be better to someone try and run my notebook. A second eye might be needed to understand the problem

**Comment by fkiraly** [comment]
hmmm, thanks. Will have a look.

**Comment by benHeid** [comment]
## Architecture Folder Proposal:
### Overview:
- forecasting.networks.ltsf contains the LTSFLinearForecaster.
  * This forecaster should implement _fit, _predict, ... 
- networks.ltsf contains LTSFLinearNetwork:
  * This object contains the neural network 
  * Challenge is that this LTSFLinearNetwork have to inherit from torch.nn.module. This raises since we have to avoid top-level imports.
 
 ### Network architecture:
 Wrap the networks in a get_network method. 
```python
def  get_ltsf_linear_network(**params):
    from torch import nn
    class LTSFLinearNetwork(nn.module):
        def __init__(...):
            ....
        def forward(...):
           ....

    return LSTFLinearNetwork(**param)
``` 

### Forecaster architecture:
I would propose to just implement _fit, _predict, etc. methods. 

```python
class LTSFLinearForecaster(BaseTorchForecaster)
    def __init__(...):
        ....
     
    def fit(...):
        ...
        self. network = get_ltsf_linear_network(**params) 
       # params might also include parameters derived from the forecasting horizon or 
       #      from provided data, e.g. dimensionality of the exogenous features. 
        ...
        self.network.fit(...)
        ....

    # Further methods. 
    
class BaseTorchForecaster(BaseForecaster):
    class SpecificTorchdataset(....):
         # Do we need this?
         # Not sure if this is the correct place? Any comments on that? 
         
    def _create_dataset(...):
        # create torch Dataset

   def _get_dataloader(...):
       # pass the dataset to the dataloader
``` 

## Questions:
Input on this architecture are highly appreciated. Especially on the following topics:
* Makes this architecture sense? 
* Where should we locate the definition of the TorchDataset? 
   * Probably the default `TensorDataset` is sufficient in most case

**Comment by fkiraly** [comment]
> Makes this architecture sense?

Yes, absolutely! Great design imo!

The one small change I would make - for discussion - is have a `BaseObject` child class, e.g., `BasePytorchNet` and concrete descendants instead of `get_ltsf_linear_network`.

Why? Because then we can use these as parametric components in estimators where we can switch out networks.

**Comment by fkiraly** [comment]
Personally, I think having the older model separately, e.g., for comparison or reference, might be useful.

**Comment by geetu040** [comment]
There are many available implementations for these models, they are mostly alike but differ slightly, I can either duplicate a specific one or study the differences and opt a unified approach for all models. I think second option suits more but if there is a specific one that only I should use, let me know

Here are the implementations that I found
- https://github.com/MAZiqing/FEDformer/tree/master/models
- https://github.com/thuml/Autoformer/tree/main/models
- https://github.com/cure-lab/LTSF-Linear/tree/main/models
- https://github.com/zhouhaoyi/Informer2020/tree/main/models
- https://github.com/ant-research/Pyraformer/tree/master/pyraformer
- https://github.com/MAZiqing/FEDformer/tree/master/models

**Comment by fkiraly** [comment]
I believe the implementations by `cure-lab` are the most popular ones because they exist in one repo.

For scientific accuracy, the *first* implementation, optimally by paper authors or affiliates, should be the reference to base things on. It might be slightly invovled to get the copy-paste trail right, i.e., which implementation was first and who is the actual author/implementer. There could be cases where the first implementer is not the author and later there is an implementation by the author, etc. The root problem seems to be that no one has released their reference implementation on pypi, and academics kept copy-pasting code around.

Doing this book-keeping once and reporting it here would be great.
I.e., for each model, list sources, and whether it is the earliest implementation, and whether it is an implementation by the original paper authors or affiliates (which need not be the same).

I am aware of at least two versions of most of the above, and three of FEDformer.
